### ISSUE TITLE
GCM forgets about projects in a failed state

### ISSUE THREAD

**Issue by wyb1** [gardener/gardener#6402]
**How to categorize this issue?**

/area robustness
/kind bug

**What happened**:

The GCM does not reconcile projects that are in a failed state if it initially gets a too many requests error. Perhaps other errors will cause the same issue, but I have only seen it with this error.

```
Events:
  Type     Reason                    Age   From                         Message
  ----     ------                    ----  ----                         -------
  Warning  NamespaceReconcileFailed  11h   gardener-controller-manager  Error while reconciling RBAC resources for namespace "garden-project": 1 error occurred:
           * the server has received too many requests and has asked us to try again later (patch rolebindings.rbac.authorization.k8s.io gardener.cloud:system:project-viewer)
```
Somehow the GCM forgets about these projects and will only reconcile if it gets restarted.

**What you expected to happen**:

GCM should eventually reconcile the failed projects.

**Comment by rfranzke** [comment]
Can you please provide some steps for reproducing this locally? This makes it easier to look into it.

**Comment by BeckerMax** [comment]
Seems like Wesley is on vacation now. If this happens again we will gather more information/ping you.

**Comment by rfranzke** [comment]
Any updates? :-)

**Comment by rfranzke** [comment]
@ialidzhikov suggested to introduce a mock error at the respective step and try to reproduce it this way. I will try doing so and see what I can find out.

**Comment by rfranzke** [comment]
I spent some time with it and added the following patch introducing a mock error:

```diff
diff --git a/pkg/operation/botanist/component/projectrbac/projectrbac.go b/pkg/operation/botanist/component/projectrbac/projectrbac.go
index 66939c239..758e64b71 100644
--- a/pkg/operation/botanist/component/projectrbac/projectrbac.go
+++ b/pkg/operation/botanist/component/projectrbac/projectrbac.go
@@ -71,6 +71,7 @@ type projectRBAC struct {
 }

 func (p *projectRBAC) Deploy(ctx context.Context) error {
+       return fmt.Errorf("dummy")
        var (
                admins                 []rbacv1.Subject
                members                []rbacv1.Subject
```

The `gardener-controller-manager` produced the following output:

```
2022-08-24T16:48:19.407+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:19.420+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:19.437+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:19.466+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:19.511+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:19.599+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:19.767+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:20.096+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:20.744+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:22.031+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:24.599+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:29.730+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:48:39.989+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:49:00.478+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:49:41.449+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:51:03.380+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:53:47.237+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T16:59:14.933+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T17:10:10.313+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T17:26:50.341+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
2022-08-24T17:43:30.389+0200	ERROR	controller.project.project	Error reconciling request	{"name": "rf", "namespace": "", "error": "dummy"}
```

We can observe the expected behavior with exponentially backing off the reconciliation up to the maximum time of `16m`.
Apart from this, I cannot see any obvious problem with the implementation or any concept misuse which could cause the reported behavior.

TL;DR: I don't think that we have a problem with the controller itself. My suspicion would rather be an broken watch connection or some bug in the underlying client library packages which might not have re-established the watch.

**Comment by rfranzke** [comment]
It's not clear how to proceed here, so let's
/close
Feel free to `/reopen` with additional information that helps to track down the issue.
